statistics/cmsketch.go

```diff
@@ -116,37 +116,43 @@ func NewCMSketchWithTopN(d, w int32, sample [][]byte, numTop uint32, rowCount ui
 	// In some cases, if user triggers fast analyze when rowCount is close to sampleSize, unexpected bahavior might happen.
 	rowCount = mathutil.MaxUint64(rowCount, uint64(len(sample)))
 	estimateNDV, scaleRatio := calculateEstimateNDV(helper, rowCount)
-	c := buildCMSWithTopN(helper, d, w, scaleRatio)
-	c.calculateDefaultVal(helper, estimateNDV, scaleRatio, rowCount)
+	defaultVal := calculateDefaultVal(helper, estimateNDV, scaleRatio, rowCount)
+	c := buildCMSWithTopN(helper, d, w, scaleRatio, defaultVal)
 	return c, estimateNDV, scaleRatio
 }
 
-func buildCMSWithTopN(helper *topNHelper, d, w int32, scaleRatio uint64) (c *CMSketch) {
+func buildCMSWithTopN(helper *topNHelper, d, w int32, scaleRatio uint64, defaultVal uint64) (c *CMSketch) {
 	c = NewCMSketch(d, w)
 	enableTopN := helper.sampleSize/topNThreshold <= helper.sumTopN
 	if enableTopN {
 		c.topN = make(map[uint64][]*TopNMeta)
 	}
+	c.defaultValue = defaultVal
 	for counterKey, cnt := range helper.counter {
-		data, scaledCount := hack.Slice(string(counterKey)), cnt*scaleRatio
+		data := hack.Slice(string(counterKey))
+		// If the value only occurred once in the sample, we assumes that there is no difference with
+		// value that does not occurred in the sample.
+		rowCount := defaultVal
+		if cnt > 1 {
+			rowCount = cnt * scaleRatio
+		}
 		if enableTopN && cnt >= helper.lastVal {
 			h1, h2 := murmur3.Sum128(data)
-			c.topN[h1] = append(c.topN[h1], &TopNMeta{h2, data, scaledCount})
+			c.topN[h1] = append(c.topN[h1], &TopNMeta{h2, data, rowCount})
 		} else {
-			c.insertBytesByCount(data, scaledCount)
+			c.insertBytesByCount(data, rowCount)
 		}
 	}
 	return
 }
 
-func (c *CMSketch) calculateDefaultVal(helper *topNHelper, estimateNDV, scaleRatio, rowCount uint64) {
```

Comment:
Any reason why change this method?

Reply:
We need the default value before constructing cm sketch, so change it to return values.